### PR TITLE
EPMLSTRCMW-209 feat: Add support for MongoDB in integration tests

### PR DESCRIPTION
**Pull request by gominasian**
**Description:**

Added support for Mongo DB in integration tests

**Changes:**

- Added TestMongoEngine
- Added support for Mongo DB test container
- Added test case which uses Mongo DB
- Added required dependency

### CODE REVIEW

File: utils/src/test/scala/cromwell/pipeline/utils/TestContainersUtils.scala [hunk 1]
```diff
@@ -27,16 +25,19 @@ object TestContainersUtils {
...
-    config("database.mongo.port" -> container.getMappedPort(mongoPort))
+  def getMongoContainer(mongoImageName: String = "mongo"): MongoDBContainer =
+    MongoDBContainer(DockerImageName.parse(mongoImageName))
+
+  def getConfigForMongoContainer(container: MongoDBContainer): MongoConfig =
+    MongoConfig(
+      mongoConfig.getString("user"),
+      mongoConfig.getString("password").toCharArray,
+      mongoConfig.getString("host"),
+      container.mappedPort(mongoPort),
```
Comment (Nikolai-Isaev) [inline utils/src/test/scala/cromwell/pipeline/utils/TestContainersUtils.scala:36]: Why do we need to treat port so distinctively here?

File: utils/src/test/scala/cromwell/pipeline/utils/TestContainersUtils.scala [hunk 2]
```diff
@@ -27,16 +25,19 @@ object TestContainersUtils {
...
-  def getMongoContainer(mongoImageName: String = "mongo"): MongoContainer =
-    new MongoContainer(mongoImageName)
-      .withExposedPorts(mongoPort)
-      .withCommand("--replSet", "docker-rs")
-      .waitingFor(Wait.forLogMessage(".*Waiting for connections.*", 1))
-
-  def getConfigForMongoContainer(container: GenericContainer[_]): Config =
-    config("database.mongo.port" -> container.getMappedPort(mongoPort))
+  def getMongoContainer(mongoImageName: String = "mongo"): MongoDBContainer =
+    MongoDBContainer(DockerImageName.parse(mongoImageName))
```
Comment (DruzhininPavel) [inline utils/src/test/scala/cromwell/pipeline/utils/TestContainersUtils.scala:29]: Where are you using this method ?

**Review by DruzhininPavel — COMMENTED** [review]
Looks good. But one question is...

File: utils/src/test/scala/cromwell/pipeline/utils/TestContainersUtils.scala (hunk 2 above)
Comment (gominasian) [reply utils/src/test/scala/cromwell/pipeline/utils/TestContainersUtils.scala:29]: It's obsolete. It comes from the previous solution. I'll fix it

**Review by DruzhininPavel — APPROVED** [review]
Nice work!

**Review by Nikolai-Isaev — APPROVED** [review]
Great job

File: datasource/src/main/scala/cromwell/pipeline/database/TestMongoEngine.scala [hunk 3]
```diff
@@ -0,0 +1,16 @@
+package cromwell.pipeline.database
+
+import cromwell.pipeline.utils.MongoConfig
+import org.mongodb.scala.{ MongoClientSettings, ServerAddress }
+
+import scala.collection.JavaConverters._
+
+class TestMongoEngine(mongoConfig: MongoConfig) extends MongoEngine(mongoConfig) {
```
Comment (myazinn) [inline datasource/src/main/scala/cromwell/pipeline/database/TestMongoEngine.scala]: Why is `TestMongoEngine` not in `test` directory?

File: repositories/src/test/scala/cromwell/pipeline/datastorage/dao/mongo/DocumentRepositoryTest.scala [hunk 4]
```diff
@@ -1,17 +1,61 @@
...
 
-class DocumentRepositoryTest extends AsyncWordSpec with Matchers with MockitoSugar {
+import java.nio.file.Paths
+import scala.concurrent.Await
+import scala.concurrent.duration.Duration
 
-  private val updateResult = mock[UpdateResult]
-  private val collection = mock[MongoCollection[Document]]
+class DocumentRepositoryTest
+    extends WordSpec
```
Comment (myazinn) [inline repositories/src/test/scala/cromwell/pipeline/datastorage/dao/mongo/DocumentRepositoryTest.scala]: Why did you change `AsyncWordSpec` to `WordSpec`? All tests here are using `Future[Assertion]`

File: repositories/src/test/scala/cromwell/pipeline/datastorage/dao/mongo/DocumentRepositoryTest.scala [hunk 5]
```diff
@@ -1,17 +1,61 @@
...
+    with Matchers
+    with MockitoSugar
+    with ScalatestRouteTest
```
Comment (myazinn) [inline repositories/src/test/scala/cromwell/pipeline/datastorage/dao/mongo/DocumentRepositoryTest.scala]: Why do we need `ScalatestRouteTest` ?

File: utils/src/test/scala/cromwell/pipeline/utils/TestContainersUtils.scala [hunk 6]
```diff
@@ -27,16 +25,16 @@ object TestContainersUtils {
...
-    new MongoContainer(mongoImageName)
-      .withExposedPorts(mongoPort)
-      .withCommand("--replSet", "docker-rs")
-      .waitingFor(Wait.forLogMessage(".*Waiting for connections.*", 1))
-
-  def getConfigForMongoContainer(container: GenericContainer[_]): Config =
-    config("database.mongo.port" -> container.getMappedPort(mongoPort))
+  def getConfigForMongoContainer(container: MongoDBContainer): MongoConfig =
+    MongoConfig(
+      mongoConfig.getString("user"),
```
Comment (myazinn) [inline utils/src/test/scala/cromwell/pipeline/utils/TestContainersUtils.scala]: Please use named arguments here

File: repositories/src/test/scala/cromwell/pipeline/datastorage/dao/mongo/DocumentRepositoryTest.scala [hunk 7]
```diff
@@ -1,17 +1,61 @@
...
+    with TestContainerForAll {
+
+  override val containerDef: MongoDBContainer.Def = MongoDBContainer.Def()
+  val container: MongoDBContainer = startContainers()
```
Comment (myazinn) [inline repositories/src/test/scala/cromwell/pipeline/datastorage/dao/mongo/DocumentRepositoryTest.scala]: Can't we make it look similar to what we have for Postgres container?

File: utils/src/test/scala/cromwell/pipeline/utils/TestContainersUtils.scala [hunk 8]
```diff
@@ -27,16 +25,16 @@ object TestContainersUtils {
   private lazy val mongoConfig = ConfigFactory.load().getConfig("database.mongo")
   private def mongoPort: Int = mongoConfig.getInt("port")
 
-  class MongoContainer(dockerImageName: String) extends GenericContainer[MongoContainer](dockerImageName)
-
-  def getMongoContainer(mongoImageName: String = "mongo"): MongoContainer =
```
Comment (myazinn) [inline utils/src/test/scala/cromwell/pipeline/utils/TestContainersUtils.scala:32]: Why did you remove this method? It's nice to have one method that gives mongo container as we did with postgres

File: repositories/src/test/scala/cromwell/pipeline/datastorage/dao/mongo/DocumentRepositoryTest.scala [hunk 9]
```diff
@@ -1,17 +1,61 @@
...
+
+  lazy implicit val config: MongoConfig = TestContainersUtils.getConfigForMongoContainer(container)
+  lazy val mongoCollection: MongoCollection[Document] =
+    new TestMongoEngine(config).mongoCollection
+  lazy val documentRepository = new DocumentRepository(mongoCollection)
+  lazy val mongoRepository: ProjectConfigurationRepository = ProjectConfigurationRepository(documentRepository)
 
-  private val documentRepository = new DocumentRepository(collection)
+  val dummyProject: Project = TestProjectUtils.getDummyProject()
+  val projectId: ProjectId = dummyProject.projectId
```
Comment (myazinn) [inline repositories/src/test/scala/cromwell/pipeline/datastorage/dao/mongo/DocumentRepositoryTest.scala]:
```suggestion
  val projectId: ProjectId = TestProjectUtils.getDummyProjectId
```

File: datasource/src/main/resources/application.conf [hunk 10]
```diff
@@ -14,4 +14,21 @@ database {
   liquibase {
     changeLogResourcePath = "liquibase/changelog/changelog-master.xml"
   }
+
+  mongo {
```
Comment (myazinn) [inline datasource/src/main/resources/application.conf:18]: I wonder why we need it here. We already have configurations for mongo in `portal` module, so why duplicate them here? I assume there is a reason for it, so I'm just asking

**Review by myazinn — CHANGES_REQUESTED** [review]

File: datasource/src/main/scala/cromwell/pipeline/database/TestMongoEngine.scala (hunk 3 above)
Comment (gominasian) [reply datasource/src/main/scala/cromwell/pipeline/database/TestMongoEngine.scala]: Fixed

File: repositories/src/test/scala/cromwell/pipeline/datastorage/dao/mongo/DocumentRepositoryTest.scala (hunk 4 above)
Comment (gominasian) [reply repositories/src/test/scala/cromwell/pipeline/datastorage/dao/mongo/DocumentRepositoryTest.scala]: Fixed

File: repositories/src/test/scala/cromwell/pipeline/datastorage/dao/mongo/DocumentRepositoryTest.scala (hunk 5 above)
Comment (gominasian) [reply repositories/src/test/scala/cromwell/pipeline/datastorage/dao/mongo/DocumentRepositoryTest.scala]: I get ExecutionContext there. Yep, not the best solution. Fixed

File: utils/src/test/scala/cromwell/pipeline/utils/TestContainersUtils.scala (hunk 6 above)
Comment (gominasian) [reply utils/src/test/scala/cromwell/pipeline/utils/TestContainersUtils.scala]: Fixed

File: repositories/src/test/scala/cromwell/pipeline/datastorage/dao/mongo/DocumentRepositoryTest.scala (hunk 7 above)
Comment (gominasian) [reply repositories/src/test/scala/cromwell/pipeline/datastorage/dao/mongo/DocumentRepositoryTest.scala]: Fixed

File: utils/src/test/scala/cromwell/pipeline/utils/TestContainersUtils.scala (hunk 8 above)
Comment (gominasian) [reply utils/src/test/scala/cromwell/pipeline/utils/TestContainersUtils.scala:32]: I thought it's redundant, but I've changed my code and it's useful now

File: repositories/src/test/scala/cromwell/pipeline/datastorage/dao/mongo/DocumentRepositoryTest.scala (hunk 9 above)
Comment (gominasian) [reply repositories/src/test/scala/cromwell/pipeline/datastorage/dao/mongo/DocumentRepositoryTest.scala]: Have cleaned it up

File: datasource/src/main/resources/application.conf (hunk 10 above)
Comment (gominasian) [reply datasource/src/main/resources/application.conf:18]: We need it in repository tests, somehow it takes config from here

File: utils/src/test/scala/cromwell/pipeline/utils/TestContainersUtils.scala [hunk 11]
```diff
@@ -27,16 +25,22 @@ object TestContainersUtils {
...
-    new MongoContainer(mongoImageName)
-      .withExposedPorts(mongoPort)
-      .withCommand("--replSet", "docker-rs")
-      .waitingFor(Wait.forLogMessage(".*Waiting for connections.*", 1))
-
-  def getConfigForMongoContainer(container: GenericContainer[_]): Config =
-    config("database.mongo.port" -> container.getMappedPort(mongoPort))
+  def getMongoContainer(mongoImageName: String = "mongo"): MongoDBContainer = {
+    val container = MongoDBContainer(DockerImageName.parse(mongoImageName))
+    container.start()
```
Comment (myazinn) [inline utils/src/test/scala/cromwell/pipeline/utils/TestContainersUtils.scala]: Why does method `getMongoContainer` starts it? Who will stop this container once tests are completed?
This method should only return the container, starting it is up to caller

File: repositories/src/test/scala/cromwell/pipeline/datastorage/dao/mongo/DocumentRepositoryTest.scala [hunk 12]
```diff
@@ -1,17 +1,44 @@
...
+import scala.concurrent.duration.Duration
+import scala.concurrent.{ Await, ExecutionContext }
+
 class DocumentRepositoryTest extends AsyncWordSpec with Matchers with MockitoSugar {
 
-  private val updateResult = mock[UpdateResult]
-  private val collection = mock[MongoCollection[Document]]
+  val container: MongoDBContainer = getMongoContainer()
+
+  override implicit val executionContext: ExecutionContext = ExecutionContext.global
```
Comment (myazinn) [inline repositories/src/test/scala/cromwell/pipeline/datastorage/dao/mongo/DocumentRepositoryTest.scala]: Why do you override test execution context with the global one?

File: repositories/src/test/scala/cromwell/pipeline/datastorage/dao/mongo/DocumentRepositoryTest.scala [hunk 13]
```diff
@@ -1,17 +1,44 @@
...
+  lazy val mongoRepository: ProjectConfigurationRepository = ProjectConfigurationRepository(documentRepository)
+
+  val projectConfiguration: ProjectConfiguration = getDummyProjectConfiguration()
 
-  private val documentRepository = new DocumentRepository(collection)
+  "Mongo repository getById" should {
+    "return some configuration" in {
+      Await.result(mongoRepository.updateConfiguration(projectConfiguration).flatMap { _ =>
+        mongoRepository.getById(projectConfiguration.id)
+      }, Duration.Inf) shouldBe Some(projectConfiguration)
```
Comment (myazinn) [inline repositories/src/test/scala/cromwell/pipeline/datastorage/dao/mongo/DocumentRepositoryTest.scala]: You are using `AsyncWordSpec`, you **don't** need to await for the result. Please read about the difference between `AsyncWordSpec` and `WordSpec`

**Review by myazinn — CHANGES_REQUESTED** [review]

File: repositories/src/test/scala/cromwell/pipeline/datastorage/dao/mongo/DocumentRepositoryTest.scala (hunk 13 above)
Comment (gominasian) [reply repositories/src/test/scala/cromwell/pipeline/datastorage/dao/mongo/DocumentRepositoryTest.scala]: Have read it and got the difference, thank you! Fixed

File: repositories/src/test/scala/cromwell/pipeline/datastorage/dao/mongo/DocumentRepositoryTest.scala (hunk 12 above)
Comment (gominasian) [reply repositories/src/test/scala/cromwell/pipeline/datastorage/dao/mongo/DocumentRepositoryTest.scala]: There was a problem with lock in first test but it's resolved now

File: utils/src/test/scala/cromwell/pipeline/utils/TestContainersUtils.scala (hunk 11 above)
Comment (gominasian) [reply utils/src/test/scala/cromwell/pipeline/utils/TestContainersUtils.scala]: Fixed

**Review by myazinn — APPROVED** [review]

File: datasource/src/main/resources/application.conf [hunk 14]
```diff
@@ -14,4 +14,21 @@ database {
...
+    user = ${?CMWL_MONGO_USER}
+    user = "user"
```
Comment (Dmitrii-Shevchenko) [inline datasource/src/main/resources/application.conf:20]: You don't need to specify the fields manually, they are already specified in the .env.example

File: datasource/src/test/scala/cromwell/pipeline/database/TestMongoEngine.scala [hunk 15]
```diff
@@ -0,0 +1,16 @@
...
+import org.mongodb.scala.{ MongoClientSettings, ServerAddress }
+
+import scala.collection.JavaConverters._
+
+class TestMongoEngine(mongoConfig: MongoConfig) extends MongoEngine(mongoConfig) {
+  override lazy val mongoSettings: MongoClientSettings = MongoClientSettings
+    .builder()
+    .applyToSslSettings(p => p.enabled(false))
+    .applyToClusterSettings(
+      p => p.hosts(List(ServerAddress("localhost", mongoConfig.port)).asJava)
```
Comment (Dmitrii-Shevchenko) [inline datasource/src/test/scala/cromwell/pipeline/database/TestMongoEngine.scala]: What is "p"? Rename it to "builder" or something else

Comment (gominasian) [reply datasource/src/test/scala/cromwell/pipeline/database/TestMongoEngine.scala]: Fixed

File: datasource/src/main/resources/application.conf (hunk 14 above)
Comment (gominasian) [reply datasource/src/main/resources/application.conf:20]: Repository tests use this file

Comment (Dmitrii-Shevchenko) [reply datasource/src/main/resources/application.conf:20]: If you have ```user = ${?CMWL_MONGO_USER}``` why do you specify ```user = "user"``` after?

**Review by Dmitrii-Shevchenko — APPROVED** [review]
